tensorflowonspark/pipeline.py

```diff
@@ -0,0 +1,313 @@
+  def getExportDir(self):
+    return self.getOrDefault(self.export_dir)
+
+class HasMethodName(Params):
+  method_name = Param(Params._dummy(), "method_name", "Method name for a saved_model signature", typeConverter=TypeConverters.toString)
+  def __init__(self):
+    super(HasMethodName, self).__init__()
+  def setMethodName(self, value):
+    return self._set(method_name=value)
+  def getMethodName(self):
+    return self.getOrDefault(self.method_name)
+
+class HasSignatureDefKey(Params):
+  signature_def_key = Param(Params._dummy(), "signature_def_key", "Identifier for a specific saved_model signature", typeConverter=TypeConverters.toString)
+  def __init__(self):
+    super(HasSignatureDefKey, self).__init__()
+  def setSignatureDefKey(self, value):
+    return self._set(signature_def_key=value)
+  def getSignatureDefKey(self):
+    return self.getOrDefault(self.signature_def_key)
+
+class HasTagSet(Params):
+  tag_set = Param(Params._dummy(), "tag_set", "Comma-delimited list of tags identifying a saved_model metagraph", typeConverter=TypeConverters.toString)
+  def __init__(self):
+    super(HasTagSet, self).__init__()
+  def setTagSet(self, value):
+    return self._set(tag_set=value)
+  def getTagSet(self):
+    return self.getOrDefault(self.tag_set)
+
+class Namespace(object):
+  """
+  Utility class to convert dictionaries to Namespace-like objects
+  Based on https://docs.python.org/dev/library/types.html#types.SimpleNamespace
+  """
+  def __init__(self, d):
+    self.__dict__.update(d)
+  def __repr__(self):
+    keys = sorted(self.__dict__)
+    items = ("{}={!r}".format(k, self.__dict__[k]) for k in keys)
+    return "{}({})".format(type(self).__name__, ", ".join(items))
+  def __eq__(self, other):
+    return self.__dict__ == other.__dict__
+
+class TFParams(Params):
+  """Mix-in class to store args and merge params"""
+  args = None
+  def _merge_args_params(self):
+    local_args = copy.copy(self.args)
+    args_dict = vars(local_args)
+    for p in self.params:
+      args_dict[p.name] = self.getOrDefault(p.name)
+    return local_args
+
+class TFEstimator(Estimator, TFParams, HasInputCol, HasPredictionCol,
+                  HasInputTensor, HasOutputTensor,
+                  HasClusterSize, HasNumPS, HasRDMA, HasTensorboard, HasModelDir,
+                  HasBatchSize, HasEpochs, HasSteps,
+                  HasExportDir, HasMethodName, HasSignatureDefKey, HasTagSet):
+  """Spark ML Pipeline Estimator which launches a TensorFlowOnSpark cluster for training"""
+
+  train_fn = None
+
+  def __init__(self, train_fn, tf_args):
+    super(TFEstimator, self).__init__()
+    self.train_fn = train_fn
+    self.args = Namespace(tf_args) if isinstance(tf_args, dict) else tf_args
+    self._setDefault(inputCol='images',
+                    predictionCol='prediction',
+                    tensor_in='input',
+                    tensor_out='output',
+                    cluster_size=1,
+                    num_ps=0,
```

Comment:
Yeah, I was trying to default to one node, e.g. `cluster_size=1, num_ps=0` implies just one `worker`.  This is mostly to force folks who want distributed to explicitly set these args in a configuration that makes sense for their app.

As for the verification, there is an assert inside TFoS, but we should probably add something inside the Estimator as well.  Note: I didn't spend much time on validations, since we were in prototyping mode.